File: python/799. Champagne Tower.py
class Solution(object):
    def champagneTower(self, poured, query_row, query_glass):
        """
        :type poured: int
        :type query_row: int
        :type query_glass: int
        :rtype: float
        """
        dp = [[0]*100 for _ in range(100)]
        dp[0][0] = poured
        for i in range(1, 100):
            count = 0
            for j in range(0, i+1):
                ans = 0
                if j > 0:
                    ans += (dp[i-1][j-1]-1)/2.0 if dp[i-1][j-1] > 1 else 0
                if j < i:
                    ans += (dp[i-1][j]-1)/2.0 if dp[i-1][j] > 1 else 0
                dp[i][j] = ans
                if ans > 1: # 本层有杯子溢出
                    count += 1
            if count==0: # 本层没有杯子溢出，跳出循环
                break
        return min(1, dp[query_row][query_glass])
# 思路: dp
# time: O(N^2), space: O(N^2)

File: go/src/leetcode/799/799. Champagne Tower.go
/**
@author ZhengHao Lou
Date    2022/11/20
*/
package main

import "fmt"

/**
https://leetcode.cn/problems/champagne-tower/
*/
const N = 100

func champagneTower(poured int, query_row int, query_glass int) float64 {
	if query_row == 0 {
		return min(1, float64(poured))
	}
	var last = []float64{float64(poured)}
	for i := 1; i < N; i++ {
		var tmp = make([]float64, i+1)
		for j := 0; j <= i; j++ {
			if j-1 >= 0 && last[j-1] > 1 {
				tmp[j] += (last[j-1] - 1) * 0.5
			}
			if j <= i-1 && last[j] > 1 {
				tmp[j] += (last[j] - 1) * 0.5
			}

			if i == query_row && j == query_glass {
				return min(1, tmp[j])
			}
		}
		last = tmp
	}

	return 0
}

func min(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

func main() {
	fmt.Println(champagneTower(4, 2, 0))
}
